src/continuous_agg.c

```diff
@@ -259,59 +358,165 @@ ts_continuous_agg_get_all_caggs_info(int32 raw_hypertable_id)
 
 	all_caggs_info.bucket_widths = NIL;
 	all_caggs_info.max_bucket_widths = NIL;
 	all_caggs_info.mat_hypertable_ids = NIL;
+	all_caggs_info.bucket_functions = NIL;
+
 	Assert(list_length(caggs) > 0);
 
 	foreach (lc, caggs)
 	{
 		ContinuousAgg *cagg = lfirst(lc);
 
-		bucket_width = Int64GetDatum(ts_continuous_agg_bucket_width(cagg));
+		bucket_width = Int64GetDatum(ts_continuous_agg_bucket_width_variable(cagg) ?
+										 BUCKET_WIDTH_VARIABLE :
+										 ts_continuous_agg_bucket_width(cagg));
 		all_caggs_info.bucket_widths =
 			lappend(all_caggs_info.bucket_widths, DatumGetPointer(bucket_width));
 
-		bucket_width = Int64GetDatum(ts_continuous_agg_max_bucket_width(cagg));
+		bucket_width = Int64GetDatum(ts_continuous_agg_bucket_width_variable(cagg) ?
+										 BUCKET_WIDTH_VARIABLE :
+										 ts_continuous_agg_max_bucket_width(cagg));
 		all_caggs_info.max_bucket_widths =
 			lappend(all_caggs_info.max_bucket_widths, DatumGetPointer(bucket_width));
 
+		bucket_function = palloc(sizeof(*bucket_function));
+		*bucket_function = cagg->bucket_function;
+		all_caggs_info.bucket_functions = lappend(all_caggs_info.bucket_functions, bucket_function);
+
 		all_caggs_info.mat_hypertable_ids =
 			lappend_int(all_caggs_info.mat_hypertable_ids, cagg->data.mat_hypertable_id);
 	}
 	return all_caggs_info;
 }
 
+/*
+ * Serializes ContinuousAggsBucketFunction* into a palloc'ated string like:
+ *
+ *     ver;bucket_width;origin;timezone
+ *
+ * ... where ver is a version of the serialization format. This particular format
+ * was chosen because of it's simplicity and good performance. Future versions
+ * of the procedure can use other format (like key=value or JSON) and/or add
+ * extra fields. NULL pointer is serialized to an empty string.
+ *
+ * Note that the schema and the name of the function are not serialized. This
+ * is intentional since this information is currently not used for anything.
+ * We can serialize the name of the function as well when and if this would be
+ * necessary.
+ */
+static const char *
+bucket_function_serialize(const ContinuousAggsBucketFunction *bf)
+{
+	const int ver = 1;
+	StringInfo str;
+
+	if (bf == NULL)
+		return pstrdup("");
+
+	str = makeStringInfo();
+
+	/* We are pretty sure that user can't place ';' character in these fields */
+	Assert(strstr(bf->bucket_width, ";") == NULL);
+	Assert(strstr(bf->origin, ";") == NULL);
+	Assert(strstr(bf->timezone, ";") == NULL);
+
+	appendStringInfo(str, "%d;%s;%s;%s;", ver, bf->bucket_width, bf->origin, bf->timezone);
+
+	return str->data;
+}
+
+/*
+ * Deserielizes a string into a palloc'ated ContinuousAggsBucketFunction*. Note
+ * that NULL is also a valid return value.
+ *
+ * See bucket_function_serialize() for more details.
+ */
+static ContinuousAggsBucketFunction *
+bucket_function_deserialize(const char *str)
+{
+	int i;
+	char *begin, *end, *strings[4];
+	ContinuousAggsBucketFunction *bf;
+
+	/* empty string stands for serialized NULL */
+	if (*str == '\0')
```

Comment:
We could allow for `const char *str` to be `NULL` since we return `NULL`. This way we won't need all those arrays and structures be populated.

Reply:
Not sure if I understand you. Could you please clarify which arrays and structures you propose not to populate and how allowing str == NULL will help with this?

Reply:
I'm marking this as resolved 🤷‍♂️